### PR TITLE
[sparse] add ftrl optimizer for sparse

### PR DESCRIPTION
**Pull request by CNevd**
@eric-haibin-lin

### CODE REVIEW

**Comment by eric-haibin-lin** [comment]
Thanks for the contribution! I'll take a look after completing some documentation work.

File: src/operator/optimizer_op.cc [hunk 1]
```diff
@@ -259,5 +260,30 @@ to be 0.9 and the learning rate :math:`\eta` to be 0.0001.
 .add_argument("delta", "NDArray-or-Symbol", "delta")
 .add_arguments(RMSPropAlexParam::__FIELDS__());
 
+NNVM_REGISTER_OP(ftrl_update)
+.describe(R"code(Update function for Ftrl optimizer.
+Referenced from *Ad Click Prediction: a View from the Trenches*, available at
+http://dl.acm.org/citation.cfm?id=2488200.
+
+.. math::
+  \\eta_{t,i} = \\frac{learningrate}{\\beta+\\sqrt{\\sum_{s=1}^tg_{s,i}^t}}
```
Comment (eric-haibin-lin) [inline src/operator/optimizer_op.cc]: Could you explicitly state the sparse update rule like the one here in the documentation?
https://github.com/apache/incubator-mxnet/blob/master/src/operator/optimizer_op.cc#L159

Comment (CNevd) [reply src/operator/optimizer_op.cc]: pls check again

File: python/mxnet/optimizer.py [hunk 2]
```diff
@@ -798,9 +820,6 @@ class Ftrl(Optimizer):
         The initial learning rate.
     beta : float, optional
         Per-coordinate learning rate correlation parameter.
-    eta :
-        .. math::
```
Comment (eric-haibin-lin) [inline python/mxnet/optimizer.py]: We should keep the clarification for eta chosen for this updater. Showing it as one of the parameters is not right place, probably just move it to line 791

Comment (eric-haibin-lin) [reply python/mxnet/optimizer.py]: Should the last part of the math be `sqrt{\\sum_{s=1}^tg_{s,i}^2}}` instead?  @moveforever

File: python/mxnet/optimizer.py [hunk 3]
```diff
@@ -790,6 +789,29 @@ class Ftrl(Optimizer):
...
+        z += rescaled_grad - (sqrt(n + rescaled_grad**2) - sqrt(n)) * weight / learning_rate
+        n += rescaled_grad**2
+        w = (sign(z) * lamda1 - z) / ((beta + sqrt(n)) / learning_rate + wd) * (abs(z) > lamda1)
+
+    If the storage types of weight, state and grad are all ``row_sparse``, \
+    sparse updates are applied by::
+
+        for row in grad.indices:
+            rescaled_grad[row] = clip(grad[row] * rescale_grad, clip_gradient)
+            z[row] += rescaled_grad[row] - (sqrt(n[row] + rescaled_grad[row]**2) - \
```
Comment (eric-haibin-lin) [inline python/mxnet/optimizer.py]: backslash produces a problem displaying these in the documentation page on mxnet.io. Please keep it in one line and add ` # pylint: disable=line-too-long` before this class and ` # pylint: disable=line-too-long` after the class

File: src/operator/optimizer_op-inl.h [hunk 4]
```diff
@@ -1035,6 +1035,213 @@ inline void RMSPropUpdate(const nnvm::NodeAttrs &attrs, const OpContext &ctx,
...
+  MSHADOW_REAL_TYPE_SWITCH(weight.type_flag_, DType, {
+    MSHADOW_IDX_TYPE_SWITCH(grad.aux_type(kIdx), IType, {
+      MXNET_ASSIGN_REQ_SWITCH(req, req_type, {
+        const DType* weight_data = weight.dptr<DType>();
+        const IType* grad_idx = grad.aux_data(kIdx).dptr<IType>();
+        const DType* grad_val = grad.data().dptr<DType>();
+        DType* z_data = z.dptr<DType>();
+        DType* n_data = n.dptr<DType>();
+        DType* out_data = out->dptr<DType>();
+        nnvm::dim_t num_rows = grad.aux_shape(kIdx)[0];
```
Comment (eric-haibin-lin) [inline src/operator/optimizer_op-inl.h]: nit: const

File: src/operator/optimizer_op-inl.h [hunk 5]
```diff
@@ -1035,6 +1035,213 @@ inline void RMSPropUpdate(const nnvm::NodeAttrs &attrs, const OpContext &ctx,
...
+                                    const NDArray& n,
+                                    const OpReqType& req,
+                                    NDArray *out) {
+  using namespace mshadow;
+  using namespace mshadow::expr;
+  using namespace mxnet_op;
+  using namespace rowsparse;
+  CHECK_RSP_ALL_ROWS_NON_ZERO(weight, "FtrlUpdate", "weights");
+  Stream<xpu>* s = ctx.get_stream<xpu>();
+  // fill mean and variance with zero values in order to reuse the sgd mom dns impl
```
Comment (eric-haibin-lin) [inline src/operator/optimizer_op-inl.h]: pls update the comment accordingly for ftrl

File: python/mxnet/optimizer.py (hunk 2 above)
Comment (moveforever) [reply python/mxnet/optimizer.py]: yes, it is. The eta is actually per-coordinate learning rate,. The equation should be moved to line 791

File: src/operator/optimizer_op.cu [hunk 6]
```diff
@@ -51,5 +51,9 @@ NNVM_REGISTER_OP(rmsprop_update)
 NNVM_REGISTER_OP(rmspropalex_update)
 .set_attr<FCompute>("FCompute<gpu>", RMSPropAlexUpdate<gpu>);
 
+NNVM_REGISTER_OP(ftrl_update)
+.set_attr<FCompute>("FCompute<gpu>", FtrlUpdate<gpu>)
+.set_attr<FComputeEx>("FComputeEx<gpu>", FtrlUpdateEx<gpu>);
```
[the remark this answers is not in the document]
Comment (eric-haibin-lin) [reply src/operator/optimizer_op.cu]: Are you suggesting removing `FCompute<gpu>` for Ftrl op? If we remove `FtrlUpdate<gpu>` from the op registration it will break backward compatibility and affect other people who are using it.

[the remark this answers is not in the document]
Comment (moveforever) [reply src/operator/optimizer_op.cu]: oh, i know. FCompute\<gpu\> is for default storage. FComputeEx is for non-default storage. tks.

File: python/mxnet/optimizer.py [hunk 7]
```diff
@@ -24,10 +24,9 @@
 import logging
 import warnings
 import numpy
-from .base import py_str
```
Comment (eric-haibin-lin) [inline python/mxnet/optimizer.py]: Please keep this line. `py_str` is needed.
